### DISCUSSION TITLE
Blackpill one pin can't be set

### DISCUSSION THREAD

**Discussion by mogplus88** [micropython/micropython#15629] [STM32 / Pyboard]
I have a problem with one pin, A11, on my Blackpill. When I run the command:
`x = Pin('A11', Pin.IN, Pin.PULL_UP)`
Thonny crashes with this error:
`PROBLEM IN THONNY'S BACK-END: Exception while handling 'execute_source' (ConnectionError: device reports readiness to read but returned no data (device disconnected or multiple access on port?)).`
To get the board working again I have to press the reset button on the board, then reconnect with Thonny.
If I run `x = Pin('A11')` it works fine, and then displaying x returns `Pin(Pin.cpu.A11, mode=Pin.ALT, alt=10)`
I can even define the pin with Pin.ALT and alt=10, but nothing else works. 
This is the only pin that has this problem, all the others seem to work fine (so far...) including A10 and A12.
Is it just a faulty pin that I can never use, or is there a way to correct this? I'm suspecting the former...
All advice gratefully received.

**Comment by davefes** [comment]
Which firmware are you running?  Maybe, a link to your `mpconfigboard.h` might help.

I have only used:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython

**Reply by mogplus88** [reply]
@davefes thanks for the quick reply! The short answer is I don't know for sure but it appears to be either pyboard or possibly a variation of pyboard, as I can import pyb and use the functions in it.
I had a look at the link you posted and it looks like the answer to a maiden's prayer (as my dear old mum used to say). I have tried on several occasions to build micropython for STM from source using various sets of instructions on the net, but it always fails. I will try one of the versions on the link you provided, I'm just not sure which version of blackpill I have. I have another one the same, and so far it has resisted all my attempts to load mp on it, or even work at all (tried it on Arduino ide). A complete dud maybe?...
I tried to attach a copy of mpconfig.h (renamed to mpconfig.h.txt so I could attach it), but there doesn't seem to be any indicator that it worked. I might have attached it twice...

**Reply by scruss** [reply]
when you hit Ctrl-D in the REPL, what message do you get back? (It starts with "MicroPython v1.23. …"

What chip is on your Black Pill?

Do you have the one with external flash? If so, there may  be some pins permanently affected by the flash connection

**Comment by davefes** [comment]
I have built firmware using his instructions, when we were trying to sort out some USB issues.  Where did you get your boards from, be aware that there are clones?  Is there a chip on the back of the board?

You can do a build for the smallest flash and use that on any board ... of course NOT getting access to the added flash.  I had to do that on boards with faulty added flash.

He has been very helpful in the past sorting out issues.

**Reply by davefes** [reply]
It did not attach.

**Reply by davefes** [reply]
How do you load the firmware?  I use STM32CubeProgrammer-2.14.0  For file loading I use `rshell` and pins A9 and A10.

**Comment by mogplus88** [comment]
@davefes @scruss thanks for your replies. This is what I get with a ctrl-D:

`MicroPython v1.22.0-preview.42.g2590a34ed on 2023-10-21; WeActStudio BlackPill v3.1 with STM32F411CE`

I never got as far as actually attempting to flash the board as the build always failed before that. Of course I can't find the instructions I followed now, but it was the last step that failed due to missing libraries. I checked the folders where the libraries were supposed to be, and they were there, so I don't know why it couldn't find them. Anyway if there's a prebuilt version I'm more than happy to use that. I'm not techie enough to be able to sort out compile and build problems. I can follow instructions, but that's about the limit of my expertise.  :(

Yes I suspect my boards are probably clones. I bought them a while back and they've been languishing in my box of bits. I don't know how mp was installed on one of them, I don't recall doing it.  Maybe it came preflashed?

The chip has "stm32f411ceu6g?23211v? chn gq 224" written on it. The question marks are characters I couldn't decipher, even with a magnifying glass. There is no chip soldered to the underneath.

I have an ST-Link programmer but I have never used it. I had a couple of chips with stm chips on them but no usb plug, but never got around to actually trying to program the boards. I guess I'll be doing to reading on how to work it! I have found that, if I hit the reset button on the board, I can open it with Dolphin and see all my python files, and I can add and delete files too.

Thanks again for your help, much appreciated.

**Reply by davefes** [reply]
V1.22 is fairly recent, like a year old?

Looks like you or someone possibly used the instructions at 
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython
or
https://github.com/jkorte-dev/micropython-board-STM32F411CE/tree/main/WEACT_F411CE 

My 8MB version is:
MicroPython v1.22.0-preview.162.g6117aa686.dirty on 2023-11-19; WeActStudio BlackPill v3.1 with STM32F411CE
but I may have built that myself.

As you do not have added flash try this one:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython/tree/main/firmware/WEACTF411CE
I use STM32Cube Programmer and STLink to get the image on.

STM32CubeProgrammer on Ubuntu.

No USB-C connector?  Have you got a link to where you bought them?

**Reply by scruss** [reply]
If the board is nicely finished (matte black top layer,not too many rough edges, thick PCB) it's likely that you have a real WeAct Studios one. They are quite well made.

You can also flash them in DFU mode, as described in Mike Causer's older repo for these boards: [Flashing via DFU](https://github.com/mcauser/WEACT_F411CEU6?tab=readme-ov-file#flashing-via-dfu). No ST-Link or large STM32Cube programmer required.

TBH, v1.22 is probably recent enough not to worry too much about. Sorry we don't have any idea why one pin is always set

**Reply by davefes** [reply]
You can flash some units in DFU mode.  Unfortunately, some 25MHz crystals are out-of-spec for use with the USB.  After discussions with WeACT they changed to a 8MHz crystal. I think they knew that there was a problem.

**Reply by scruss** [reply]
good to know. The 25 MHz ones I have seemed to work in DFU mode

**Reply by davefes** [reply]
@mogplus88, what is Dolphin?  I get hits for `Dolphin emulator`, but that doesn't look relevant.

**Reply by mogplus88** [reply]
@davefes [dolphin](https://apps.kde.org/dolphin/) is a file manager for Linux, like Explorer on Windows. Apologies for the lack of clarity, sometimes I forget that there are people out there who don't use Linux. ;-)

The boards I have with no USB connector I bought a while back because they were cheap and had lots of pins. No idea where I got them, but probably Banggood as I tend to buy stuff there on a whim, because they are the next shiny thing...

@scruss The boards I have do seem to be good quality, clear printing of labels and matt black surface. Reasonably thick pcb and quality soldering too, so maybe they are genuine. Thanks for the tips.

I'm pretty happy with the board, and one dud pin isn't going to make much difference. I do have another one the same though, and I'd like to flash mp onto that one too, and dfu seems like the easier option. However, slight problem, I tried to use df-util and got a 'command not found'. Did a quick search to find how to install it but couldn't find anything. I'm using Ultramarine linux (Fedora based) and zsh shell. What am I missing?

Thanks.

**Reply by mogplus88** [reply]
Never mind. Shoulda cleaned my glasses. It's dfu-util. No wonder I couldn't find it.

**Reply by davefes** [reply]
I do use Linux :)  Seems some people have to resort to a hairdryer or some heat on the 25MHz xtal to hopefully bring it unto frequency for dfu to work.

**Reply by davefes** [reply]
I just noticed in the README,md here:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython
and in `Pins Definition` that there is a note about some of the pins, ie A11

**Reply by mogplus88** [reply]
Hmmm... got a whole other problem now. My non-mp board seems to have died completely. Yesterday when I tried to access it with Thonny it filled the repl screen with garbage. Now Thonny can't even find the board, and neither can `ls /dev/tty*`. I followed the instructions to put it in dfu mode, which seem to have worked as dfu-util found it, but returned the error:
`dfu-util: Cannot open DFU device 0483:df11 found on devnum 10 (LIBUSB_ERROR_ACCESS)`
If it was heavier I could use it as a paper weight. Looks like it's a candidate for the bin, unless there is a way to forcibly flash it or something. 

:-(

**Reply by mogplus88** [reply]
@davefes that's interesting. Maybe it's not an error, just the pin can't be used? Unless it's used for USART6 or SPI4? Dunno...

**Reply by davefes** [reply]
That is might be the way it is set it up.  Would need to know the actual firmware on the device.

Try holding your finger on the 25MHz crystal for a minute or so.  Seems the assumption is that if the crystal is out of spec then all or most are either on the high side or low side of 25MHz.  Don't know what their temperature coefficient is.

**Reply by mogplus88** [reply]
@davefes @scruss, don't know what the heck happened, but...

I tried to rebuild mp from scratch, and again it failed with the weirdass library missing even though it isn't. So, just for fun, tried it on another distro I have installed, Debian based this time, MX. Again the build failed. So, because I'm a glutton for punishment, I tried it on another one, good ole faithful Mint. I nearly fell of my chair when the build worked. 

So, started fiddling with board that doesn't work and dfu. After plugging and unplugging numerous times, and putting it in dfu mode, hit `dfu-util -l` and it worked! Holy crap! Tried to flash it with dfu, and it worked first time!!! Can't believe it. Bought myself a lottery ticket on the strength of it. 

The board now seems to be working fine. Interestingly pin A11, and A12 as it turns out, can't be defined as anything but their defaults, so it must be as has been mentioned earlier, that the pins are perhaps dedicated to another function and can't be used for anything else. c'est la vie. I can live with that.

Once again, thanks so much to all who have contributed to this discussion, much appreciated.

:-)

**Reply by davefes** [reply]
Which board are you building for?

**Reply by mogplus88** [reply]
@davefes it was for the Blackpill I have that wasn't working.
For the record, this is the first few lines of the error output I got when I tried to run the build (on Ultramarine and MX). I realised as I was doing this that the file `objgetitemiter.c` exists, that's what I checked for the first time I tried it, but I now realise it was actually complaining about the included lib `stdlib.h`. You'd think with all that underlining I would have noticed it, but no... I did a search for that but I couldn't find it. There were a couple of them in the micropython folder, but they were in the pico-sdk library, so not relevant to stm32 I'm assuming. I don't know where include libraries are supposed to be...

`Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
GEN build-WEACT_F411CEU6/genhdr/qstr.i.last
../../py/objgetitemiter.c:27:10: fatal error: stdlib.h: No such file or directory
   27 | #include <stdlib.h>
      |          ^~~~~~~~~~
compilation terminated.
In file included from ../../py/mpstate.h:29,
                 from ../../py/mpstate.c:27:
/usr/lib/gcc/arm-none-eabi/14.1.0/include/stdint.h:9:16: fatal error: stdint.h: No such file or directory
    9 | # include_next <stdint.h>
      |                ^~~~~~~~~~
compilation terminated.
In file included from timer.c:27:
/usr/lib/gcc/arm-none-eabi/14.1.0/include/stdint.h:9:16: fatal error: stdint.h: No such file or directory
    9 | # include_next <stdint.h>
      |                ^~~~~~~~~~
compilation terminated.
In file included from ../../py/binary.h:30,
                 from ../../extmod/modbluetooth.c:28:
../../py/obj.h:29:10: fatal error: assert.h: No such file or directory
   29 | #include <assert.h>
      |          ^~~~~~~~~~
`

**Reply by davefes** [reply]
Whose board definition are you using?  In `micropython/ports/stm32/boards` you must be selecting a specific board.  Or, whose instructions are you following?

I suspect that it was a problem with your previous build environment.

**Reply by mogplus88** [reply]
I used the one you suggested, the nspsck one.

**Reply by davefes** [reply]
OK.  Now you get start making some projects.

**Comment by mogplus88** [comment]
Yeah, or rebuilding in my case. I have a couple of ancient and venerable radio control transmitters circa late 70's early 80's that I converted to 2.4GHZ with FrSky Tx modules and home brewed encoders. One using an Arduino Nano and the other using a Teensy LC. Tried them recently and neither of them worked, not sure why. The Teensy LC isn't suported any more, and the Nano was getting a bit long in the tooth, so I thought I'd replace them with something that could run micropython. I had a couple of Blackpills lying around in my parts box, so they were pressed into service. Just about got the coding done, but the hard bit is all the soldering. I also have a couple of more complex Tx's which I'm probably going to put ESP32's into, and maybe 900MHz modules, but that's a long term project...
;-)

**Comment by andrewleech** [comment]
On a related note I've pushed up a new MR for a board profile for this board: https://github.com/micropython/micropython/pull/15646
It's not reinventing anything, just taking previous definitions and bringing them up to current codebase requirements.

The schematic for the Black Pill v2.0 can be found here: https://stm32-base.org/assets/pdf/boards/original-schematic-STM32F411CEU6_WeAct_Black_Pill_V2.0.pdf

Of note, and indeed on any STM32F411, Pins A11 and A12 are the USB interface pins! 
So if they're reconfigured it's expected that the USB interface would drop out, I'm surprised @mogplus88 that you could do anything with A12 without also killing the USB :-D

**Reply by davefes** [reply]
@andrewleech Is the MR for the public?  If so, where do I find it?  Does this handle the 25MHz and 8MHz variants?  This project seems to keep up to date. https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython

**Reply by andrewleech** [reply]
Hi @davefes the link to my pr is in my previous comment: "#15646"

Currently it's just covering a couple of 25mhz variants, though I could certainly add more if they make sense. 

I wasn't aware of anyone else keeping up to date builds for this, I'll take a look thanks.

**Reply by davefes** [reply]
I had gone to "#15646" but could not work-out where to go after that.  The new 8MHz variant gets-around issues with using DFU.

**Reply by andrewleech** [reply]
The PR is a submission for inclusion here, if it's accepted then builds will be available of the official micropython download page. There's no public builds from it until then however. 

Yeah I can understand the benefit of the 8mhz boards, though I can't see any mention of them on the official GitHub https://github.com/WeActStudio/WeActStudio.MiniSTM32F4x1

Oh, I can see the mention of it on their official AliExpress : https://a.aliexpress.com/_mKC3jxW

Disappointingly it looks like they still call the 8mhz version v3.1 the same as the current 25mhz. That makes it harder for users to know what they've got! 
Regardless, yeah that would be an easy variant to add.

**Reply by davefes** [reply]
OK, got it.

You would have to look closely at the marking on the crystal on the board.

**Reply by robert-hh** [reply]
At Aliexpress they sell both the 8M and 25M version for a marginally different price. A few cent less for the 8M version.

**Reply by davefes** [reply]
Got a negative response from WeACT if they planned to make a 8MHZ 8MB variant, so I bought some 8MHz SMD crystal units.  Still haven't tried swapping them out.

**Reply by robert-hh** [reply]
If the 8MHz version are only available without Flash, then it requires only a single variant.
It look as if they all will be sold next week for 99Ct including shipping!

**Reply by davefes** [reply]
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython/tree/main/firmware has been kind enough to provide firmware for 3 variants using 8MHz for those of us brave enough to put more flash on the 8MHz boards.

Thanks for the heads-up, I would like a few more.

**Comment by mogplus88** [comment]
@andrewleech  you are correct, A12 does exactly the same thing. When I was first trying to track down this problem I only got as far as 
a11. But trying to redefine A12 also causes Thonny to crash. Hardly surprising if it's fouling up the usb port... 
Live and learn.
;-)
